File: Loja - Java/src/sporthub/Estoque.java
package sporthub;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.List;


/**
 *
 * @author Weslles
 */
public class Estoque {
    private List <Equipamento> listaEquipamento = new ArrayList<>();
    private List <Vestuario> listaVestuario= new ArrayList<>();

    public void insereListaProdutos(String cpf){
        Scanner input = new Scanner(System.in);
        int marca, qnt, option;
        float preco;

        
        System.out.println("\n---------------\n1-Peça de Vestuário\n2-Equipamento Esportivo\n---------------");
        System.out.println("\nSelecione a opcao que melhor especifica o produto a ser inserido::");
        option=input.nextInt();
        while (option < 1 || option > 2){
            System.out.println("Opcao inexistente! informe uma opcao valida.");
            option=input.nextInt();
        }
        
        System.out.println("Selecione a opcao que indica a marca do produto: ");
        System.out.println("\n1 - Adidas"
                         + "\n2 - Fila"
                         + "\n3 - Nike"
                         + "\n4 - Olympikus"
                         + "\n5 - Oxer"
                         + "\n6 - Puma"
                         + "\n7 - Topper");
        marca = input.nextInt();
        while ( marca < 1 || marca > 7 ){
            System.out.println("Opcao inexistente! informe uma opcao valida.");
            marca = input.nextInt();
        }
        
        System.out.println("\nInforme a quantidade de produtos que deseja inserir: ");
        qnt = input.nextInt();
        System.out.println("Informe o preço do produto: ");
        preco = input.nextFloat();
        
        switch(option){
            case 1:{
                insereListaVestuario(marca, cpf, qnt, preco);
                break;
            }
            case 2:{
                insereListaEquipamento(marca, cpf, qnt, preco);
                break;
            }
        }
    } // OK
   
    public void retiraListaProdutos(String cpf){
        int qnt, option, id;
        Scanner ler = new Scanner(System.in);
        System.out.println("\nSelecione a opcao que melhor especifica o produto a ser retirado:");
        System.out.println("1 - Peça de Vestuário");
        System.out.println("2 - Equipamento Esportivo\n");
        option = ler.nextInt();
        System.out.println("\n Informe o ID do produto que deseja retirar: ");
        System.out.println("Se não souber, pressione 0 para voltar e verifique na checagem do estoque");
        id = ler.nextInt();
        
        if( id == 0)
            return;
        System.out.println("\nInforme quantos produtos devem ser retirados: ");
        qnt = ler.nextInt();
        switch(option){
            case 1:{
                retiraListaVestuario(qnt, id, cpf);
                break;
            }
            case 2:{
                retiraListaEquipamento(qnt, id, cpf);
                break;
            }
        }
    }  //OK  
    
    public void insereListaEquipamento(int marca, String cpf, int qnt, float preco ){
        int aux1, aux2, desc, mod, ind =0;
        int i;
        
        Scanner ler = new Scanner(System.in);
       
        System.out.println("Selecione a opção que melhor descreve o produto: ");
        System.out.println("\n1 - Bicicleta"
                + "\n2 - Bola"
                + "\n3 - Patins"
                + "\n4 - Prancha"
                + "\n5 - Raquete"
                + "\n6 - Rede");
        desc = ler.nextInt();
        while ( desc < 1 || desc >6 ){
            System.out.println("Opcao inexistente! informe uma opcao valida.");
            desc = ler.nextInt();
        }
        
        System.out.println("Selecione a opção que indica a modalidade esportiva do equipamento: ");
        System.out.println("\n1 - Basquete"
                + "\n2 - Ciclismo"
                + "\n3 - Futebol"
                + "\n4 - Patinação"
                + "\n5 - Surf"
                + "\n6 - Tenis"
                + "\n7 - Volei");
        mod = ler.nextInt();
        while ( mod < 1 || mod >6 ){
            System.out.println("Opcao inexistente! informe uma opcao valida.");
            mod = ler.nextInt();
        }
        
        Equipamento e = new Equipamento(marca, desc, cpf, qnt, 2, preco, mod);
        while (ind < listaEquipamento.size()){
            if(     listaEquipamento.get(ind).getMarca().equals(e.getMarca()) &&
                    listaEquipamento.get(ind).getDescricao().equals(e.getDescricao()) &&
                    listaEquipamento.get(ind).getPreco() == e.getPreco() &&
                    listaEquipamento.get(ind).getModalidade().equals(e.getModalidade()) ){
                aux1 = listaEquipamento.get(ind).getQuantidade();
                aux2 = e.getQuantidade();
                listaEquipamento.get(ind).setQuantidade(aux1+aux2);
                listaEquipamento.get(ind).setCpf(cpf);
                break;
            }
            ind++; 
        }
        
        if(ind == listaEquipamento.size()){
            if(ind == 0)
                i = 1;
            else{
                ind--;
                i = listaEquipamento.get(ind).getId()+1;
            }
                
            e.setId(i);
            listaEquipamento.add(e);
        }
    } // OK

    public void insereListaVestuario(int marca, String cpf, int quantidade, float preco){
        Scanner ler = new Scanner(System.in);
        int desc, tam, genero, ind = 0;
        int aux1, aux2;
        
        System.out.println("Selecione a opcao que melhor descreve o produto:");
        System.out.println(  "1 - Calça\n"
                           + "2 - Calçado\n"
                           + "3 - Camiseta\n"
                           + "4 - Meia\n"
                           + "5 - Shorts\n");
        
        desc = ler.nextInt();
        while( desc < 1 || desc > 5){
            System.out.println("Opcao inexistente! informe uma opcao valida.");
            desc = ler.nextInt();
        }
        System.out.println("Informe o tamanho da(s) peças: ");
        tam = ler.nextInt();
        System.out.println("Selecione a opção que indica o genero do produto: ");
        System.out.println("1 - Feminino"+
                         "\n2 - Masculino");
        
        genero = ler.nextInt();
        while ( genero < 1 || genero > 2){
            System.out.println("Opcao inexistente! informe uma opcao valida.");
            genero = ler.nextInt();
        }
        Vestuario v = new Vestuario(marca, desc, cpf, quantidade, 1, tam, genero, preco);
        while(ind < listaVestuario.size() ){
            if( listaVestuario.get(ind).getGenero().equals(v.getGenero())&&
                listaVestuario.get(ind).getDescricao().equals(v.getDescricao()) &&
                listaVestuario.get(ind).getMarca().equals(v.getMarca()) &&
                listaVestuario.get(ind).getPreco() == v.getPreco() &&
                listaVestuario.get(ind).getTamanho() == v.getTamanho() ){
                
                    aux1 = listaVestuario.get(ind).getQuantidade();
                    aux2 = v.getQuantidade();
                    listaVestuario.get(ind).setQuantidade(aux1+aux2);
                    listaVestuario.get(ind).setCpf(v.getCpf());
                    break;
            }
            ind++;
        }
        if ( ind == listaVestuario.size() ){
            int novoInd;
            if( ind == 0 )
                novoInd = 1;
            else{
                ind--;
                novoInd = listaVestuario.get(ind).getId()+1;
            }
            v.setId(novoInd);
            listaVestuario.add(v);
        }
        
    }

    public void retiraListaEquipamento(int quantidade, int id, String cpf){
        int cont, ind = 0, cheqId = 0;
        boolean b = false;
                
        while( cheqId != id && ind <= listaEquipamento.size() ){
            cheqId = listaEquipamento.get(ind).getId();
            if(cheqId == id){
                cont = listaEquipamento.get(ind).getQuantidade();
                if(cont < quantidade)
                    System.out.println("Erro! Não existem tantos exemplares no estoque.");
                else{
                    listaEquipamento.get(ind).setQuantidade(cont - quantidade);
                    listaEquipamento.get(ind).setCpf(cpf);
                }
                b = true;
                break;
            }
            ind++;
                 
        }
        if( b == false){
            System.out.println("Erro! O produto nao se encontra disponivel no estoque");
        }
    }  //OK

    public void retiraListaVestuario(int quantidade, int id, String cpf ){
        int cont, ind = 0, cheqId = 0;
        boolean b = false;
                
        while( cheqId != id && ind <= listaVestuario.size() ){
            cheqId = listaVestuario.get(ind).getId();
            if(cheqId == id){
                cont = listaVestuario.get(ind).getQuantidade();
                if(cont < quantidade)
                    System.out.println("Erro! Não existem tantos exemplares no estoque.");
                else{
                    listaVestuario.get(ind).setQuantidade(cont - quantidade);
                    listaVestuario.get(ind).setCpf(cpf);
                }      
                b = true;
                break;
            }
            ind++;
                 
        }
        if( b == false){
            System.out.println("Erro! O produto nao se encontra disponivel no estoque");
        }
    }  //OK

    public void imprimeListaVestuario() {
        int i;
        for ( i = 0; i < listaVestuario.size(); i++){
            listaVestuario.get(i).imprimeInformacoesVestuario();
        }
     }  //OK

    public void imprimeListaEquipamento() {
        int j;
        for(j = 0; j<listaEquipamento.size(); j++){
            listaEquipamento.get(j).imprimeInformacoesEquipamento();
        }
     }   //OK

    public void checaEstoque(){
        int op;
        boolean b = true;
        int v;
        char t;
        Scanner ler = new Scanner(System.in);

        System.out.println("Informe o tipo de produto que deseja verificar: ");
        System.out.println("1 - Produtos de Vestuário ");
        System.out.println("2 - Equipamentos Esportivos ");
        System.out.println("3 - Todo o Estoque ");
        op = ler.nextInt();

        switch(op){
            case 1:{
                System.out.println("Lista de produtos de vestuario disponiveis no estoque: ");
                imprimeListaVestuario();
                break;
            }
            case 2:{
                System.out.println("Lista de Equipamentos presentes no estoque: ");
                imprimeListaEquipamento();
                break;
            }
            case 3:{
                System.out.println("Lista de Produtos de Vestuario disponivel no estoque: ");
                imprimeListaVestuario();
                System.out.println("\nLista de Equipamentos presentes no estoque: ");
                imprimeListaEquipamento();
                break;
            }
            default:
                System.out.println("Opção Invalida!");
            
        }
        System.out.println("\nPara voltar digite 0: ");
        v = ler.nextInt();
        while(v != 0 ){
            System.out.println("\nErro! Para voltar digite v: ");
            v = ler.nextInt();
        }
    } 
}

File: Loja - C++/Estoque.cpp
#include "Estoque.h"

using namespace std;

void Estoque::insereListaProdutos(string cpf)
{
    int marca, quantidade, option;
    float preco;
    cout << "Selecione a opcao que melhor especifica o produto a ser inserido: "<< endl;
    cout << "1 - Peca de vestuario" << endl;
    cout << "2 - Equipamento esportivo" << endl;
    fflush(stdin);
    cin >> option;
    while (option < 1 || option > 2)
    {
        cout << "Opcao inexistente! informe uma opcao valida." << endl;
        fflush(stdin);
        cin >> option;
    }
    cout << "Selecione a opcao que indica a marca do produto: " << endl;
    cout << "1 - Adidas" << endl;
    cout << "2 - Fila" << endl;
    cout << "3 - Nike" << endl;
    cout << "4 - Olympikus" << endl;
    cout << "5 - Oxer" << endl;
    cout << "6 - Puma" << endl;
    cout << "7 - Topper" << endl;
    fflush(stdin);
    cin >> marca;
    while (marca < 1 || marca > 7)
    {
        cout << "Opcao inexistente! informe uma opcao valida." << endl;
        fflush(stdin);
        cin >> marca;
    }
    cout << "Informe a quantidade de produtos que deseja inserir: " << endl;
    fflush(stdin);
    cin >> quantidade;
    cout << "Informe o preco do produto: " << endl;
    fflush(stdin);
    cin >> preco;
    switch (option)
    {
    case 1:
        {
            insereListaVestuario(marca, cpf, quantidade, preco);
        }
        break;
    case 2:
        {
            insereListaEquipamento(marca, cpf, quantidade, preco);
        }
        break;
    }
}

void Estoque::retiraListaProdutos(string cpf)
{
    int quantidade, option, id;
    cout << "Selecione a opcao que melhor especifica o produto a ser retirado: "<< endl;
    cout << "1 - Peca de vestuario" << endl;
    cout << "2 - Equipamento esportivo" << endl;
    fflush(stdin);
    cin >> option;
    cout << "Informe o ID do produto que deseja retirar: " << endl;
    cout << "(Se nao souber, pressione 0 para voltar, e verifique na checagem de estoque) " << endl;
    fflush(stdin);
    cin >> id;
    if(id == 0)
    {
        system("cls");
        return;
    }
    cout << "Informe quantos produtos devem ser retirados: " << endl;
    fflush(stdin);
    cin >> quantidade;
    system("cls");
    switch (option)
    {
    case 1:
        {
            retiraListaVestuario(quantidade, id, cpf);
            break;
        }
    case 2:
        {
            retiraListaEquipamento(quantidade, id, cpf);
            break;
        }
    }
}

void Estoque::insereListaVestuario(int marca, string cpf, int quantidade, float preco)
{
    int tamanho, genero, aux1, aux2, descricao,j = 0;
    std::list<Vestuario>::iterator pV;
    cout << "Selecione a opcao que melhor descreve o produto: " << endl;
    cout << "1 - Calca" << endl;
    cout << "2 - Calcado" << endl;
    cout << "3 - Camiseta" << endl;
    cout << "4 - Meia" << endl;
    cout << "5 - Shorts" << endl;
    fflush(stdin);
    cin >> descricao;
    while (descricao < 1 || descricao > 5)
    {
        cout << "Opcao inexistente! informe uma opcao valida." << endl;
        fflush(stdin);
        cin >> descricao;
    }
    cout << "Informe o tamanho da(s) pecas" << endl;
    fflush(stdin);
    cin >> tamanho;
    cout << "Selecione a opcao que indica o genero do produto: " << endl;
    cout << "1 - Feminino" << endl;
    cout << "2 - Masculino" << endl;
    fflush(stdin);
    cin >> genero;
    while (genero < 1 || genero > 2)
    {
        cout << "Opcao inexistente! informe uma opcao valida." << endl;
        fflush(stdin);
        cin >> genero;
    }
    system("cls");
    Vestuario v(marca, descricao, cpf, quantidade, 1, tamanho, genero, preco);
    pV = listaVestuario.begin();
    while (j<listaVestuario.size())
        {
            if(pV->getMarca() == v.getMarca() && pV->getDescricao() == v.getDescricao() && pV->getPreco() == v.getPreco() && pV->getTamanho() == v.getTamanho() && pV->getGenero() == v.getGenero())
            {
                aux1 = pV->getQuantidade();
                aux2 = v.getQuantidade();
                pV->setQuantidade(aux1+aux2);
                pV->setCpf(cpf);
                break;
            }
            j++;
            pV++;
        }
        if (j == listaVestuario.size())
        {
            int i = j+1;
            v.setId(i);
            listaVestuario.push_back(v);
        }
}
void Estoque::insereListaEquipamento(int marca, string cpf, int quantidade, float preco)
{
    int aux1, aux2, descricao, modalidade, j = 0;
    std::list<Equipamento>::iterator pE;
    cout << "Selecione a opcao que melhor descreve o produto: " << endl;
    cout << "1 - Bicicleta" << endl;
    cout << "2 - Bola" << endl;
    cout << "3 - Patins" << endl;
    cout << "4 - Prancha" << endl;
    cout << "5 - Raquete" << endl;
    cout << "6 - Rede" << endl;
    fflush(stdin);
    cin >> descricao;
    while (descricao < 1 || descricao > 6)
    {
        cout << "Opcao inexistente! informe uma opcao valida." << endl;
        fflush(stdin);
        cin >> descricao;
    }
    cout << "Selecione a opcao que indica a modalidade esportiva do equipamento: " << endl;
    cout << "1 - Basquete" << endl;
    cout << "2 - Ciclismo" << endl;
    cout << "3 - Futebol" << endl;
    cout << "4 - Patinacao" << endl;
    cout << "5 - Surf" << endl;
    cout << "6 - Tenis" << endl;
    cout << "7 - Volei" << endl;
    fflush(stdin);
    cin >> modalidade;
    while (modalidade < 1 || modalidade > 7)
    {
        cout << "Opcao inexistente! informe uma opcao valida." << endl;
        fflush(stdin);
        cin >> modalidade;
    }
    system("cls");
    Equipamento e(marca, descricao, cpf, quantidade, 2, preco, modalidade);
    pE = listaEquipamento.begin();
    while (j<listaEquipamento.size())
        {
            if(pE->getMarca() == e.getMarca() && pE->getDescricao() == e.getDescricao() && pE->getPreco() == e.getPreco() && pE->getModalidade() == e.getModalidade())
            {
                aux1 = pE->getQuantidade();
                aux2 = e.getQuantidade();
                pE->setQuantidade(aux1+aux2);
                pE->setCpf(cpf);
                break;
            }
            j++;
            pE++;
        }
        if (j == listaEquipamento.size())
        {
            e.setId(j+1);
            listaEquipamento.push_back(e);
        }
}

void Estoque::retiraListaVestuario(int quantidade, int id, string cpf)
{
    int j, t = 0, k = 0;
    bool b = false;
    std::list<Vestuario>::iterator pV;
    pV = listaVestuario.begin();
    while (k != id && t <= listaVestuario.size())
    {
        k = pV->getId();
        if (k == id)
        {
            j = pV->getQuantidade();
            if (j < quantidade)
                cout << "Erro! Nao existem tantos exemplares no estoque." << endl;
            else if (j >= quantidade)
            {
                pV->setQuantidade(j-quantidade);
                pV->setCpf(cpf);
            }
            b = true;
            break;
        }
        pV++;
        t++;
    }
    if(b == false)
    {
        cout << "Erro! O produto nao se encontra disponivel no estoque" << endl;
    }
}

void Estoque::retiraListaEquipamento(int quantidade, int id, string cpf)
{
    int j, t = 0, k = 0;
    string modalidade;
    bool b = false;
    std::list<Equipamento>::iterator pE;
    pE = listaEquipamento.begin();
    while (k != id && t <= listaEquipamento.size())
    {
        k = pE->getId();
        if (k == id)
        {
            j = pE->getQuantidade();
            if (j < quantidade)
                cout << "Erro! Nao existem tantos exemplares no estoque." << endl;
            else if (j >= quantidade)
            {
                pE->setQuantidade(j-quantidade);
                pE->setCpf(cpf);
            }
            b = true;
            break;

        }
        pE++;
        t++;
    }
    if(b == false)
    {
        cout << "Erro! O produto nao se encontra disponivel no estoque" << endl;
    }
}

void Estoque::imprimeListaVestuario()
{
    std::list<Vestuario>::iterator pV;
    pV = listaVestuario.begin();
    for (int i=0; i<listaVestuario.size();i++)
        {
            pV->imprimeInformacoesVestuario();
            pV++;
        }
    return;
}

void Estoque::imprimeListaEquipamento()
{
    std::list<Equipamento>::iterator pE;
    pE = listaEquipamento.begin();
    for (int j=0; j<listaEquipamento.size();j++)
        {
            pE->imprimeInformacoesEquipamento();
            pE++;
        }
    return;
}

void Estoque::checaEstoque()
{
    int op;
    char v;
    cout << "Informe o tipo de produto que deseja verificar: " << endl;
    cout << "1 - Produtos de vestuario" << endl;
    cout << "2 - Equipamentos esportivos" << endl;
    cout << "3 - Todo o estoque" << endl;
    fflush(stdin);
    cin >> op;
    system("cls");
    switch (op)
    {
    case 1:
        {
            cout << "Lista de produtos de vestuario disponiveis no estoque: " << endl;
            imprimeListaVestuario();
            break;
        }
    case 2:
        {
            cout << "Lista de Equipamentos presentes no estoque: " << endl;
            imprimeListaEquipamento();
            break;
        }
    case 3:
        {
            cout << "Lista de produtos de vestuario disponiveis no estoque: " << endl;
            imprimeListaVestuario();
            cout << "\nLista de Equipamentos presentes no estoque: " << endl;
            imprimeListaEquipamento();
            break;
        }
    default:
        cout << "Opcao invalida!" << endl;
    }
    cout << "\nPara voltar digite v: " << endl;
    cin >> v;
    system("cls");
}
